### PR TITLE
fix: Build libzmq locally to work on all host types

### PR DESCRIPTION
**Pull request by pabera**
This PR addresses and aims to resolve the building issues associated with `libzmq` across all platforms.

Previously, the strategy of utilizing pre-built libzmq binaries within Docker environments was intended to streamline development processes. However, this approach did not achieve the desired outcome of creating a seamless and frustration-free development experience.

We are shifting back to locally building `libzmq`. This process is now consolidated within a dedicated Dockerfile, ensuring that the build is required only once. Once the build is completed, it becomes readily accessible and shareable across all Docker environments, significantly improving the development workflow.

Introduces changes
* Outsourced `libzmq` build into local Docker build. Artefacts are shared between Dockerfiles
* Changed naming convention for Docker files: `[name].Dockerfile` to `Dockerfile.[name]` (easier to find in file tree)
* Updated documentation. Made it easier to copy/paste individual commands by separating them

Should solve:
* #2305 (thanks @notapirate, I took a bit of inspiration for my solution from your work)
* #2167

Relates to:
* #2299 (We should not update to Bookworm yet for Docker just to resolve libzmq issues)

### Test
* [x] Mac
* [x] Windows (with WSL)
* [x] Linux

### CODE REVIEW

**Review by pabera — COMMENTED** [review]
I have only tested this on Mac so far. If others could help trying this out on their machine, that would be great.

File: docker/Dockerfile.libzmq [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+# Define environment variables for libzmq
+ENV ZMQ_VERSION 4.3.5
+ENV ZMQ_PREFIX /usr/local
+
+# Download, compile, and install libzmq
+RUN wget https://github.com/zeromq/libzmq/releases/download/v${ZMQ_VERSION}/zeromq-${ZMQ_VERSION}.tar.gz -O libzmq.tar.gz; \
+    tar -xzf libzmq.tar.gz; \
+    cd zeromq-${ZMQ_VERSION}; \
+    ./configure --prefix=${ZMQ_PREFIX} --enable-drafts; \
+    make && make install
```
Comment (AlvinSchiller) [inline docker/Dockerfile.libzmq:16]: As this is only to build the lib, would it be better to use a seperate installation dir where only the libzmq files will be located? This way they can be copied seperatly without other unrelated files in /usr/local.

`DESTDIR` like described in https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/future3/develop/documentation/developers/libzmq.md#2-cross-compilation-using-dockcross

File: docker/Dockerfile.jukebox [hunk 2]
```diff
@@ -37,21 +37,14 @@ ENV VIRTUAL_ENV=${INSTALLATION_PATH}/.venv
...
-    wget https://github.com/pabera/libzmq/releases/download/v${ZMQ_VERSION}/libzmq5-${ARCH}-${ZMQ_VERSION}.tar.gz -O libzmq.tar.gz; \
-    tar -xzf libzmq.tar.gz -C ${ZMQ_PREFIX}; \
-    rm -f libzmq.tar.gz;
-
-RUN export ZMQ_PREFIX=${PREFIX} && export ZMQ_DRAFT_API=1
-RUN pip install -v --no-binary pyzmq pyzmq
+# Install pyzmq Python dependency separately
+ENV ZMQ_PREFIX ${VIRTUAL_ENV}
+ENV ZMQ_DRAFT_API 1
+COPY --from=libzmq /usr/local ${ZMQ_PREFIX}
```
Comment (AlvinSchiller) [inline docker/Dockerfile.jukebox:46]: Copying must be done before the user is switched for sufficient access rights.
See #2305

File: docker/Dockerfile.libzmq (hunk 1 above)
Comment (AlvinSchiller) [inline docker/Dockerfile.libzmq:16]: Add `-j$(nproc)` option as described in https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/future3/develop/documentation/developers/libzmq.md#2-cross-compilation-using-dockcross

File: docker/Dockerfile.jukebox (hunk 2 above)
Comment (pabera) [reply docker/Dockerfile.jukebox:46]: Strange, it works perfectly for me ?!

Comment (pabera) [reply docker/Dockerfile.jukebox:46]: But I haven't tested other environments yet! :-)

File: docker/Dockerfile.libzmq (hunk 1 above)
Comment (pabera) [reply docker/Dockerfile.libzmq:16]: I like the idea.. will change it

File: docker/Dockerfile.jukebox (hunk 2 above)
Comment (AlvinSchiller) [reply docker/Dockerfile.jukebox:46]: ah ok, it seems like the COPY command is always related to `root` unless `--chown` is used.
So this differes from the report of @notapirate in #2305.

For readability it might still be better to change order :)

File: docker/Dockerfile.jukebox [hunk 3]
```diff
@@ -37,21 +37,14 @@ ENV VIRTUAL_ENV=${INSTALLATION_PATH}/.venv
...
-
-RUN [ "$(uname -m)" = "aarch64" ] && ARCH="arm64" || ARCH="$(uname -m)"; \
-    wget https://github.com/pabera/libzmq/releases/download/v${ZMQ_VERSION}/libzmq5-${ARCH}-${ZMQ_VERSION}.tar.gz -O libzmq.tar.gz; \
-    tar -xzf libzmq.tar.gz -C ${ZMQ_PREFIX}; \
-    rm -f libzmq.tar.gz;
-
-RUN export ZMQ_PREFIX=${PREFIX} && export ZMQ_DRAFT_API=1
-RUN pip install -v --no-binary pyzmq pyzmq
+# Install pyzmq Python dependency separately
+ENV ZMQ_PREFIX ${VIRTUAL_ENV}
```
[the remark this answers is not in the document]
Comment (pabera) [reply docker/Dockerfile.jukebox:44]: Agree with the virtual env. didn't see the comment. We need to keep `ZMQ_PREFIX` for now for your request to use a different path than `/usr/local`

File: docker/Dockerfile.jukebox (hunk 2 above)
Comment (pabera) [reply docker/Dockerfile.jukebox:46]: I'd like to keep it there as it belongs together as an entity

Comment (AlvinSchiller) [reply docker/Dockerfile.jukebox]: i rather meant moving the whole pyzmq setup

Comment (pabera) [reply docker/Dockerfile.jukebox]: But it can't go above the venv definition and then there is only the python requirements part left above it. I'd keep it that way, it works great.

Else I don't really get your proposal ;-)

Comment (AlvinSchiller) [reply docker/Dockerfile.jukebox]: Now that the path changed it could  ^^

File: docker/Dockerfile.jukebox (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlvinSchiller) [reply docker/Dockerfile.jukebox]: These path are not related.

What i meant was the following. In this case `ZMQ_PREFIX` could be ommited:

```
ENV ZMQ_PREFIX /usr/local
ENV ZMQ_DRAFT_API 1
COPY --from=libzmq /opt/libzmq ${ZMQ_PREFIX}
RUN pip install -v pyzmq --no-binary pyzmq
```

File: docker/Dockerfile.jukebox (hunk 2 above)
Comment (pabera) [reply docker/Dockerfile.jukebox]: The pip command needs to stay where it is, in order to be installed into the venv. If I move the copy command and the ENV definition higher up, I separate concerns (pyzmq) for the reason of merging another concern (copy)

Comment (AlvinSchiller) [reply docker/Dockerfile.jukebox]: my mistake, misunderstood the last comment about the venv. 
You are right, it's better kept this way 👍

File: docker/Dockerfile.libzmq [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+# Cleanup unnecessary files
+RUN rm -rf /zeromq-${ZMQ_VERSION} libzmq.tar.gz
```
Comment (notapirate) [inline docker/Dockerfile.libzmq:20]: Works fine on linux host. I only have a small suggestion to make the libzmq image smaller:

```suggestion
RUN rm -rf /zeromq-${ZMQ_VERSION} libzmq.tar.gz

# create final image with only the libzmq build fragments
FROM scratch
ENV ZMQ_PREFIX /opt/libzmq
COPY --from=0 ${ZMQ_PREFIX} ${ZMQ_PREFIX}
```

**Review by notapirate — CHANGES_REQUESTED** [review]

**Comment by pabera** [comment]
Test this on Windows (WSL, x86_64) successfully.

File: docker/Dockerfile.libzmq (hunk 4 above)
Comment (pabera) [reply docker/Dockerfile.libzmq:20]: Worked! Thanks

**Comment by s-martin** [comment]
@notapirate could you add a comment, with which platform/configuration it works for you?

**Comment by notapirate** [comment]
Sure, I tested this on Linux (Arch, x86_64) successfully.

**Comment by s-martin** [comment]
On Windows 11 building the images works now, but due to path errors the jukebox runs not:
`jukebox  | run_jukebox.py: error: argument -c/--conf: can't open '/root/RPi-Jukebox-RFID/shared/settings/jukebox.yaml': [Errno 2] No such file or directory: '/root/RPi-Jukebox-RFID/shared/settings/jukebox.yaml'`

I think we would need a Win specific docker-compose.yml for that

**Review by s-martin — APPROVED** [review]

**Review by notapirate — COMMENTED** [review]
@pabera you may want to double-check changes in docker.md, I think the item numbering got messed up :)

**Comment by s-martin** [comment]
@pabera i just merged #2310, if you use the new future3/develop you could remove the `markdownlint-disable` / `restore` tags from docker.md, because this is now configured globally.

**Comment by s-martin** [comment]
> @pabera you may want to double-check changes in docker.md, I think the item numbering got messed up :)

Not necessary, „markdown magic“ renders the numbering correctly.

**Comment by s-martin** [comment]
@pabera 
in line 379 and 393 are two more `<!-- markdownlint` tags, could you remove them as well?

**Comment by s-martin** [comment]
After #2312 is merged the error in the CodeQL should be gone.

**Review by s-martin — APPROVED** [review]

**Review by AlvinSchiller — APPROVED** [review]
